File: Sources/Buckett/Services/S3XML.swift
import Foundation

/// Tiny DOM-style wrapper around Foundation's XMLParser, sufficient for S3 responses.
final class XMLNode {
    let name: String
    var text: String = ""
    var children: [XMLNode] = []

    init(name: String) {
        self.name = name
    }

    /// First child with the given element name.
    subscript(_ childName: String) -> XMLNode? {
        children.first { $0.name == childName }
    }

    func all(_ childName: String) -> [XMLNode] {
        children.filter { $0.name == childName }
    }

    var trimmedText: String {
        text.trimmingCharacters(in: .whitespacesAndNewlines)
    }
}

final class XMLTree: NSObject, XMLParserDelegate {
    private var stack: [XMLNode] = []
    private var root: XMLNode?

    static func parse(_ data: Data) -> XMLNode? {
        let tree = XMLTree()
        let parser = XMLParser(data: data)
        parser.delegate = tree
        parser.parse()
        return tree.root
    }

    func parser(
        _ parser: XMLParser,
        didStartElement elementName: String,
        namespaceURI: String?,
        qualifiedName qName: String?,
        attributes attributeDict: [String: String] = [:]
    ) {
        let node = XMLNode(name: elementName)
        if let parent = stack.last {
            parent.children.append(node)
        } else {
            root = node
        }
        stack.append(node)
    }

    func parser(_ parser: XMLParser, foundCharacters string: String) {
        stack.last?.text += string
    }

    func parser(
        _ parser: XMLParser,
        didEndElement elementName: String,
        namespaceURI: String?,
        qualifiedName qName: String?
    ) {
        if !stack.isEmpty { stack.removeLast() }
    }
}

enum S3Date {
    /// ISO 8601 with fractional seconds, as used in S3 XML (e.g. 2024-01-02T03:04:05.000Z).
    private static let fractional: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return f
    }()

    private static let plain: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime]
        return f
    }()

    static func parse(_ string: String) -> Date? {
        fractional.date(from: string) ?? plain.date(from: string)
    }
}

/// Escapes text for embedding in request XML bodies (e.g. DeleteObjects keys).
func xmlEscape(_ s: String) -> String {
    s.replacingOccurrences(of: "&", with: "&amp;")
        .replacingOccurrences(of: "<", with: "&lt;")
        .replacingOccurrences(of: ">", with: "&gt;")
        .replacingOccurrences(of: "\"", with: "&quot;")
        .replacingOccurrences(of: "'", with: "&apos;")
}

File: windows/src/Buckett/Services/S3Xml.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Xml;
using System.Xml.Linq;

namespace Buckett.Services;

/// Tiny DOM-style wrapper around the XML reader, sufficient for S3 responses.
/// Element lookups ignore XML namespaces (S3 responses carry a default one).
public sealed class XmlNode
{
    public XmlNode(string name) => Name = name;

    public string Name { get; }
    public string Text { get; set; } = "";
    public List<XmlNode> Children { get; } = new();

    /// First child with the given element name.
    public XmlNode? this[string childName] =>
        Children.FirstOrDefault(child => child.Name == childName);

    public IReadOnlyList<XmlNode> All(string childName) =>
        Children.Where(child => child.Name == childName).ToList();

    public string TrimmedText => Text.Trim();

    /// Convenience for the very common "value of a child element, trimmed" lookup.
    public string? TextOf(string childName) => this[childName]?.TrimmedText;
}

public static class XmlTree
{
    public static XmlNode? Parse(byte[] data)
    {
        if (data.Length == 0) return null;
        try
        {
            using var stream = new MemoryStream(data);
            var settings = new XmlReaderSettings
            {
                DtdProcessing = DtdProcessing.Prohibit,
                XmlResolver = null,
                IgnoreComments = true,
                IgnoreProcessingInstructions = true
            };
            using var reader = XmlReader.Create(stream, settings);
            var document = XDocument.Load(reader, LoadOptions.None);
            return document.Root == null ? null : Convert(document.Root);
        }
        catch (Exception)
        {
            return null;
        }
    }

    private static XmlNode Convert(XElement element)
    {
        var node = new XmlNode(element.Name.LocalName);
        foreach (var child in element.Elements())
        {
            node.Children.Add(Convert(child));
        }
        if (!element.HasElements)
        {
            node.Text = element.Value;
        }
        return node;
    }
}

public static class S3Date
{
    /// ISO 8601 with or without fractional seconds, as used in S3 XML
    /// (e.g. 2024-01-02T03:04:05.000Z).
    public static DateTime? Parse(string? value)
    {
        if (string.IsNullOrWhiteSpace(value)) return null;
        if (DateTime.TryParse(
                value,
                CultureInfo.InvariantCulture,
                DateTimeStyles.AdjustToUniversal | DateTimeStyles.AssumeUniversal,
                out var parsed))
        {
            return DateTime.SpecifyKind(parsed, DateTimeKind.Utc);
        }
        return null;
    }
}

public static class S3Xml
{
    /// Escapes text for embedding in request XML bodies (e.g. DeleteObjects keys).
    public static string Escape(string value) =>
        value.Replace("&", "&amp;")
             .Replace("<", "&lt;")
             .Replace(">", "&gt;")
             .Replace("\"", "&quot;")
             .Replace("'", "&apos;");
}
